src/EnergyPlus/SimAirServingZones.cc

```diff
@@ -1441,107 +1441,50 @@ void InitAirLoops(EnergyPlusData &state, bool const FirstHVACIteration) // TRUE
-    int ZoneInNum;                // zone inlet index
-    int ZoneInSysIndex;           // index into CoolCtrlZoneNums or HeatCtrlZoneNums
-    int NumComponentsInSys;       // total number of components in the primary air system
-    int NumComponentsOnBranch;    // total number of components in the primary air system
-    bool FoundSupPathZoneConnect; // true if there is a valid connection between the supply air path and a zone terminal unit inlet
-    CompType CompTypeNum;         // component_type number for components on branches
-    int SupFanIndex;
-    int RetFanIndex;
-    bool FoundOASys;
-
-    // Dimension the local subcomponent arrays
+    int const numPrimaryAirSys = state.dataHVACGlobal->NumPrimaryAirSys;
 
-    auto &NumPrimaryAirSys = state.dataHVACGlobal->NumPrimaryAirSys;
-
-    bool ErrorsFound;
-    Array1D_int tmpNodeARR;
-    int nodeLoop;
-    int ZoneNum;
-
-    ErrorsFound = false;
+    bool ErrorsFound = false;
     state.dataHVACGlobal->AirLoopInit = true;
 
-    auto &PrimaryAirSystems(state.dataAirSystemsData->PrimaryAirSystems);
-    auto &AirToZoneNodeInfo(state.dataAirLoop->AirToZoneNodeInfo);
-    auto &AirLoopControlInfo(state.dataAirLoop->AirLoopControlInfo);
-
     // Do the one time initializations
     if (state.dataSimAirServingZones->InitAirLoopsOneTimeFlag) {
 
         // Figure out what zones are served by each primary air system (air loop) and
         // store the results in AirToZoneNodeInfo()%CoolCtrlZoneNums and AirToZoneNodeInfo()%HeatCtrlZoneNums
 
-        // Allocate scratch arrays for storing controlled zone numbers for each air loop.
-        state.dataSimAirServingZones->CtrlZoneNumsCool.allocate(state.dataGlobal->NumOfZones);
-        state.dataSimAirServingZones->CtrlZoneNumsHeat.allocate(state.dataGlobal->NumOfZones);
-        state.dataSimAirServingZones->ZoneInletNodesCool.allocate(state.dataGlobal->NumOfZones);
-        state.dataSimAirServingZones->ZoneInletNodesHeat.allocate(state.dataGlobal->NumOfZones);
-        state.dataSimAirServingZones->TermInletNodesCool.allocate(state.dataGlobal->NumOfZones);
-        state.dataSimAirServingZones->TermInletNodesHeat.allocate(state.dataGlobal->NumOfZones);
-        state.dataSimAirServingZones->TermUnitSizingNumsCool.allocate(state.dataGlobal->NumOfZones);
-        state.dataSimAirServingZones->TermUnitSizingNumsHeat.allocate(state.dataGlobal->NumOfZones);
+        // Temporary air loop zone data
+        struct AirloopZone
+        {
+            int ctrlZoneNum = 0;         // Controlled zone num
+            int zoneInletNode = 0;       // Zone supply inlet node
+            int termUnitInletNode = 0;   // Terminal unit inlet node
+            int termUnitSizingIndex = 0; // Terminal unit sizing index
+        };
+
+        EPVector<AirloopZone> cooledZone;
+        EPVector<AirloopZone> heatedZone;
+        size_t atuArraySize = max(static_cast<size_t>(state.dataGlobal->NumOfZones), state.dataDefineEquipment->AirDistUnit.size());
```

Comment:
Allocating to the max of number or zones or number of ATUs. Probably safe to go with just the number of ATUs, but not sure about that.
Everything else in this source file is code cleanup, scoping, referencing, etc.

Reply:
You could wind up with an array size larger than number of ATUs (via plenums and unconditioned zones), but that should not be an issue. Bigger picture is correct indexing of these arrays, which seems correct at the moment, just a little confusing. The only test I can think of, not as a requirement here but as a check if the indexing works, is to set up more zones than ATUs and put the non-ATU zone in the middle of the pack. Indexing should skip that zone. I don't see a problem with this as-is.